src/measure_theory/measure/finite_measure_weak_convergence.lean

```diff
@@ -0,0 +1,276 @@
+/-
+Copyright (c) 2021 Kalle Kytölä. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kalle Kytölä
+-/
+import measure_theory.measure.measure_space
+
+/-!
+# Weak convergence of (finite) measures
+
+This file will define the topology of weak convergence of finite measures and probability measures
+on topological spaces. The topology of weak convergence is the coarsest topology w.r.t. which
+for every bounded continuous `ℝ≥0`-valued function `f` on, the integration of `f` against the
+measure is continuous.
+
+TODOs:
+* Rename the types `finite_measures α` -> `finite_measure α` and
+  `probability_measures α` -> `probability_measure α` after PR #8831.
+* Define the topologies (the current version only defines the types) via
+  `weak_dual ℝ≥0 (α →ᵇ ℝ≥0)`.
+* Prove that an equivalent definition of the topologies is obtained using bounded continuous
+  `ℝ`-valued functions instead.
+* Include the portmanteau theorem on characterizations of weak convergence of (Borel) probability
+  measures.
+
+## Main definitions
+
+The main definitions are the types `finite_measures α` and `probability_measures α`.
+
+TODO:
+* Define the topologies on the above types.
+
+## Main results
+
+None yet.
+
+TODO:
+* Portmanteau theorem.
+
+## Notations
+
+No new notation is introduced.
+
+## Implementation notes
+
+The topology of weak convergence of finite Borel measures will be defined using a mapping from
+`finite_measures α` to `weak_dual ℝ≥0 (α →ᵇ ℝ≥0)`, inheriting the topology from there.
+
+## References
+
+* [Billingsley, *Convergence of probability measures*][billingsley1999]
+
+## Tags
+
+weak convergence of measures, finite measure, probability measure
+
+-/
+
+noncomputable theory
+open measure_theory
+open set
+open filter
+open_locale topological_space ennreal nnreal
+
+section finite_measures
+/-!
+### Finite measures
+In this section, we define the `Type` of finite measures on a given measurable space. This type
+will be equipped with the topology of weak convergence of measures when the measurable space is
+a topological space equipped with its Borel sigma-algebra.
+-/
+
+variables {α : Type} [measurable_space α]
+
+/-- Finite measures are defined as the subtype of measures that have the property of being finite
+measures (i.e., their total mass is finite). -/
+def finite_measures (α : Type*) [measurable_space α] : Type :=
```

Comment:
Does this behave well as an `abbreviation`? That way you get the coercion for free

Reply:
I'm not sure I understand yet. (...meaning of course: I'm sure I don't understand, at least yet).

Does `abbreviation` somehow create the coercion for free, if the abbreviation defines a subtype? Moreover, will that not interfere with trying to have instances such as `[topological_space (finite_measures α)]`? I was almost moving to the direction of making the definition `irreducible` (in my mind this is the opposite direction to making it an `abbreviation`) to avoid abused unfolding and especially to only have certain instances (topology) available when the type is explicitly `finite_measures α`.

I will test this suggestion to understand better!

Reply:
I tested this enough to see that at least nothing breaks that I feared might. I haven't yet tested the actual advantages.

Although I did not find any concrete problems, there is still something about using abbreviation (which will be def be a subtype of measures) to hold a topology instance that makes me a bit uneasy.

The good news is that I now at least saw that
```
abbreviation finite_measure (α : Type*) [measurable_space α] : Type* :=
{ μ : measure α // is_finite_measure μ }
```
can hold a topology instance. This wasn't obvious to me from the start. (Although then come think of it, neither should it have been obvious to me that a `def` can hold a topology instance; I was just more familiar with the latter.)

But there will actually be many natural topologies on finite measures and probability measures. Since those will require type synonyms, isn't there a risk that abbreviation easily unfolds to its defeq subtype and then gets a wrong topology?

Specifically, for `probability_measure α`, I would shamelessly advocate making the topology of weak convergence the default instance (and I was happy that Sébastien Gouëzel at least jokingly agreed <https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.238337.20Weak.20convergence/near/250407583>). But earlier this week I discussed probability teaching with a probabilist colleague, and he seemed to use other topologies (in particular total variation distance and Wasserstein distances) almost as often. So at least I believe the need to define a few other topologies on type synonyms of `probability_measure α` will arise rather soon.

The situation with topologies on finite measures is even less clear. My own feeling now is that the default topology instance on `finite_measure α` deserves to be the one coming from the total variation metric --- this matches with the dual norm, when finite Borel measures (on reasonable topological spaces) are seen as the dual of bounded continuous functions with sup-norm. So in the long run, I think that the topology of weak convergence (which will be used in where I'm currently headed) could/should be living on a type synonym of `finite_measure α`. Has `abbreviation` been used with type synonyms in this way elsewhere?

Anyway, I have found no concrete problems, and I am yet to try the actual benefits...

Reply:
Ok, I changed to `abbreviation` and removed the `val_eq_coe` lemmas. I'm not sure I foresee all the consequences down the road (meaning again: I'm sure I don't foresee...), but it should be easy to switch back if necessary.

Also I think there will at some point be need for type synonyms with different topology instances, but we will see if the current version works smoothly when the time comes.